internal/worker/concurrency_auto_scaler.go

```diff
@@ -0,0 +1,263 @@
+	tick := defaultAutoScalerUpdateTick
+	if input.Tick != 0 {
+		tick = input.Tick
+	}
+	return &ConcurrencyAutoScaler{
+		shutdownChan:           make(chan struct{}),
+		concurrency:            input.Concurrency,
+		cooldown:               input.Cooldown,
+		log:                    input.Logger,
+		scope:                  input.Scope,
+		clock:                  input.Clock,
+		updateTick:             tick,
+		enable:                 atomic.Bool{}, // initial value should be false and is only turned on from auto config hint
+		pollerInitCount:        input.Concurrency.PollerPermit.Quota(),
+		pollerMaxCount:         input.PollerMaxCount,
+		pollerMinCount:         input.PollerMinCount,
+		pollerWaitTimeInMsLog2: newRollingAverage(numberOfPollsInRollingAverage),
+		pollerPermitLastUpdate: input.Clock.Now(),
+	}
+}
+
+func (c *ConcurrencyAutoScaler) Start() {
+	c.logEvent(autoScalerEventStart)
+
+	c.wg.Add(1)
+
+	go func() {
+		defer c.wg.Done()
+		ticker := c.clock.NewTicker(c.updateTick)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-c.shutdownChan:
+				return
+			case <-ticker.Chan():
+				c.logEvent(autoScalerEventMetrics)
+				c.updatePollerPermit()
+			}
+		}
+	}()
+}
+
+func (c *ConcurrencyAutoScaler) Stop() {
+	close(c.shutdownChan)
+	c.wg.Wait()
+	c.logEvent(autoScalerEventStop)
+}
+
+// ProcessPollerHint reads the poller response hint and take actions
+// 1. update poller wait time
+// 2. enable/disable auto scaler
+func (c *ConcurrencyAutoScaler) ProcessPollerHint(hint *shared.AutoConfigHint) {
+	if hint == nil {
+		c.log.Warn("auto config hint is nil, this results in no action")
+		return
+	}
+	if hint.PollerWaitTimeInMs != nil {
+		waitTimeInMs := *hint.PollerWaitTimeInMs
+		c.pollerWaitTimeInMsLog2.Add(math.Log2(float64(waitTimeInMs + 1)))
+	}
+
+	/*
+		Atomically compare and switch the auto scaler enable flag. If auto scaler is turned off, IMMEDIATELY reset the concurrency limits.
+	*/
+	var shouldEnable bool
+	if hint.EnableAutoConfig != nil && *hint.EnableAutoConfig {
+		shouldEnable = true
+	}
+	if switched := c.enable.CompareAndSwap(!shouldEnable, shouldEnable); switched {
+		if shouldEnable {
+			c.logEvent(autoScalerEventEnable)
+		} else {
+			c.resetConcurrency()
+			c.logEvent(autoScalerEventDisable)
+		}
+	}
+}
+
+// resetConcurrency reset poller quota to the max value. This will be used for gracefully switching the auto scaler off to avoid workers stuck in the wrong state
+func (c *ConcurrencyAutoScaler) resetConcurrency() {
+	c.concurrency.PollerPermit.SetQuota(c.pollerInitCount)
+}
+
+func (c *ConcurrencyAutoScaler) logEvent(event autoScalerEvent) {
+	if c.enable.Load() {
+		c.scope.Counter("concurrency_auto_scaler.enabled").Inc(1)
+	} else {
+		c.scope.Counter("concurrency_auto_scaler.disabled").Inc(1)
+	}
+	c.scope.Gauge("poller_in_action").Update(float64(c.concurrency.PollerPermit.Count()))
+	c.scope.Gauge("poller_quota").Update(float64(c.concurrency.PollerPermit.Quota()))
```

Comment:
no complaint at all about emitting both, but: when are we expecting them to be different, aside from for extremely small amounts of time (between "poll complete" and "started the next poll")?